src/v/cluster/controller_backend.cc

```diff
@@ -1412,7 +1418,10 @@ ss::future<std::error_code> controller_backend::create_partition(
               read_replica_bucket,
               raft::with_learner_recovery_throttle::yes,
               raft::keep_snapshotted_log::no,
-              cfg->properties.remote_label);
+              cfg->properties.remote_label,
+              std::move(xst_state));
+
+            _xst_states.erase(ntp);
```

Comment:
> Can two different moves of replicas of the same partition happen concurrently?

Of the same partition - no, but of the same NTP - yes (though very-very unlikely), in case of two partitions with different log revisions (say, after a topic with the same name was re-created). As reconciliation on different shards happens completely independently, in theory it is possible that a partition with the old log revision still exists (and maybe even is moved) on the shard that still hasn't got the info that the log revision increased.

> Is there any possibility we are erasing not the same entry as we have copied from 21 lines above?

But here we are safe and the answer is no. This is guaranteed by the entry in the shard placement table with the status `hosted_status::hosted`. We verify that it is there with the ` _shard_placement.prepare_create` call above, and it acts as a "lock", guaranteeing that nobody will move any partition state to this shard while it is there (and only this shard can remove it).

Reply:
> and only this shard can remove it

what stops it from removing or replacing it from a different fibre?

Reply:
> what stops it from removing or replacing it from a different fibre?

There is not more than one reconciliation fiber per ntp on a given shard, guaranteed by the way we launch it in `controller_backend::process_delta` and `controller_backend::notify_reconciliation`.